src/mame/palm/palm.cpp

```diff
@@ -16,195 +16,442 @@
-		m_palette(*this, "palette"),
-		m_io_penx(*this, "PENX"),
-		m_io_peny(*this, "PENY"),
-		m_io_penb(*this, "PENB"),
-		m_io_portd(*this, "PORTD")
+	palm_base_state(const machine_config &mconfig, device_type type, const char *tag)
+		: driver_device(mconfig, type, tag)
+		, m_maincpu(*this, "maincpu")
+		, m_ram(*this, RAM_TAG)
+		, m_screen(*this, "screen")
+		, m_io_penx(*this, "PENX")
+		, m_io_peny(*this, "PENY")
+		, m_io_penb(*this, "PENB")
 	{ }
 
-	void palmiii(machine_config &config);
-	void pilot1k(machine_config &config);
-	void palmvx(machine_config &config);
-	void palmv(machine_config &config);
-	void palm(machine_config &config);
-	void palmpro(machine_config &config);
-	void pilot5k(machine_config &config);
-
 	DECLARE_INPUT_CHANGED_MEMBER(pen_check);
-	DECLARE_INPUT_CHANGED_MEMBER(button_check);
 
 protected:
 	virtual void machine_start() override;
 	virtual void machine_reset() override;
 
-	enum : uint8_t
-	{
-		PORTF_Y_VCCN_BIT		= 0,
-		PORTF_Y_GND_BIT			= 1,
-		PORTF_X_VCCN_BIT		= 2,
-		PORTF_X_GND_BIT			= 3,
-		PORTF_LCD_EN_BIT		= 4,
-		PORTF_LCD_VCCN_BIT		= 5,
-		PORTF_LCD_VEE_BIT		= 6,
-		PORTF_ADC_CSN_BIT		= 7,
-
-		PORTF_PEN_MASK			= 0x8f,
-		PORTF_X_MASK			= (1 << PORTF_X_VCCN_BIT) | (1 << PORTF_Y_GND_BIT),
-		PORTF_Y_MASK			= (1 << PORTF_Y_VCCN_BIT) | (1 << PORTF_X_GND_BIT),
-	};
-
-private:
 	offs_t dasm_override(std::ostream &stream, offs_t pc, const util::disasm_interface::data_buffer &opcodes, const util::disasm_interface::data_buffer &params);
-	void mem_map(address_map &map);
 
 	void flm_out(int state);
 	void llp_out(int state);
-	void lsclk_out(int state);
-	void ld_out(uint8_t data);
+	virtual void lsclk_out(int state);
+	void ld_out(u8 data);
 	void lcd_info_changed(double refresh_hz, int width, int height);
-	uint32_t screen_update(screen_device &screen, bitmap_rgb32 &bitmap, const rectangle &cliprect);
-	void init_palette(palette_device &palette) const;
-
-	void port_f_out(uint8_t data);
-	uint8_t port_c_in();
-	uint8_t port_f_in();
 
-	int spi_in();
+	virtual int spi_from_hw();
 
-	required_device<mc68328_device> m_maincpu;
+	required_device<mc68328_base_device> m_maincpu;
 	required_device<ram_device> m_ram;
 	required_device<screen_device> m_screen;
-	required_device<palette_device> m_palette;
 	required_ioport m_io_penx;
 	required_ioport m_io_peny;
 	required_ioport m_io_penb;
-	required_ioport m_io_portd;
 
-	uint8_t m_port_f_latch;
-
-	uint16_t m_spim_data;
+	u16 m_spim_data;
 
 	bitmap_rgb32 m_lcd_bitmap;
 	int m_lcd_first_line;
 	int m_lcd_line_pulse;
 	int m_lcd_shift_clk;
-	uint8_t m_lcd_data;
+	u8 m_lcd_data;
 	int m_lcd_scan_x;
 	int m_lcd_scan_y;
 
 	static const int EXTRA_ARTWORK_HEIGHT = 60;
 };
 
+class palm_state : public palm_base_state
+{
+public:
+	palm_state(const machine_config &mconfig, device_type type, const char *tag)
+		: palm_base_state(mconfig, type, tag)
+		, m_palette(*this, "palette")
+		, m_io_portd(*this, "PORTD")
+	{ }
+
+	void palm_base(machine_config &config);
```

Comment:
Is this a concrete machine configuration, or is it just to be used by other machine configuration functions?  If the latter, it should be protected.